Guard against calling `libc::exit` multiple times on Linux.

zachs18 opened the pull request:
Mitigates (but does not fix) #126600 by ensuring only one thread which calls Rust `exit` actually calls `libc::exit`, and all other callers of Rust `exit` block.

zachs18 commented:
Hmm, this will cause a deadlock if *the same thread* calls Rust `exit` in an `atexit` handler, which is not ideal. ([playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7370e1230922bcb209ec3a063b55b154) of similar code.)

tbu- commented:
You could use a atomic variable that starts at 0 and replace the 0 (in a relaxed fashion) with the thread ID that called `exit`.

If you're on another thread that calls `process::exit` while exiting is already in progress, you can just call `pause` in a loop.

If you're on the same thread that called `process::exit` and it calls `process::exit` again, you can just exit the process directly or call `abort`.

the8472 commented:
We also don't need a full lock. Once exiting has started all other threads can `thread::park` and wait for the fires to come for them.

workingjubilee commented:
that is not dead which can eternal lie
and with strange aeons even death may die

tbu- commented:
I think this also needs to protect the exit from `main` the same way.

zachs18 committed:
On `target_os = "linux"`, ensure that only one Rust thread calls `libc::exit` or returns from `main`.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9570089210/job/26384084475) [(plain)](https://github.com/rust-lang/rust/commit/916d7fc973a0435c70971389249b943ecb6b1c28/checks/26384084475/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:692973e3d937129bcbf40652eb9f2f61becf3332)
Download action repository 'actions/upload-artifact@v4' (SHA:65462800fd760344b1a7b4382951275a0abb4808)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.414   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.428 Collecting boolean-py==4.0
#13 0.432   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.445 Collecting chardet==5.1.0
---
#13 3.492 Building wheels for collected packages: reuse
#13 3.493   Building wheel for reuse (pyproject.toml): started
#13 3.820   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.821   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 3.821   Stored in directory: /tmp/pip-ephem-wheel-cache-0elh4ajs/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 3.824 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 3.846   Attempting uninstall: setuptools
#13 3.847     Found existing installation: setuptools 59.6.0
#13 3.848     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.848     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.849     Can't uninstall 'setuptools'. No files were found to uninstall.
#13 4.512 Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
#13 4.512 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 5.017 Collecting virtualenv
#13 5.058   Downloading virtualenv-20.26.2-py3-none-any.whl (3.9 MB)
#13 5.148      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.9/3.9 MB 44.8 MB/s eta 0:00:00
#13 5.185 Collecting distlib<1,>=0.3.7
#13 5.189   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.228 Collecting filelock<4,>=3.12.2
#13 5.233   Downloading filelock-3.15.1-py3-none-any.whl (15 kB)
#13 5.233   Downloading filelock-3.15.1-py3-none-any.whl (15 kB)
#13 5.262 Collecting platformdirs<5,>=3.9.1
#13 5.266   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 5.354 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.518 Successfully installed distlib-0.3.8 filelock-3.15.1 platformdirs-4.2.2 virtualenv-20.26.2
#13 DONE 5.6s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      188352 kB
DirectMap2M:     8200192 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished `dev` profile [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/8814b926f49bc5780753ed9533853679a1181357/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-8814b926f49bc5780753ed9533853679a1181357-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
fmt check
fmt: checked 5108 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/library/std/src/rt.rs:166: platform-specific cfg: cfg(unix)
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.0)
some tidy checks failed
Build completed unsuccessfully in 0:01:08
  local time: Tue Jun 18 18:17:43 UTC 2024
  network time: Tue, 18 Jun 2024 18:17:43 GMT
```

</details>

ChrisDenton commented:
You could just define `unique_thread_exit` in `pal` and do the `cfg` there.

the8472 commented:
> thread::current() is not usable in some situations (experimentally, it fails when called from an atexit handler), so currently unique_thread_exit just aborts if thread::try_current() returns an error.

Does  `libc::gettid`  work?

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9607317914/job/26498279260) [(plain)](https://github.com/rust-lang/rust/commit/5c676be217c53553e8e3a8efa7e6d38b355a01ad/checks/26498279260/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    |                                         ^^^^^^ help: a function with a similar name exists: `getgid`
    |
   ::: /cargo/registry/src/index.crates.io-6f17d22bba15001f/libc-0.2.155/src/unix/mod.rs:880:5
    |
880 |     pub fn getgid() -> gid_t;
    |     ------------------------ similarly named function `getgid` defined here
For more information about this error, try `rustc --explain E0425`.
error: could not compile `std` (lib) due to 1 previous error
Build completed unsuccessfully in 0:01:11
  local time: Fri Jun 21 02:30:18 UTC 2024
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9607511978/job/26498795643) [(plain)](https://github.com/rust-lang/rust/commit/e71d06be1079db18c39b7088414380d94ce5a8f4/checks/26498795643/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
  --> library/std/src/sys/pal/common/exit_guard.rs:3:9
   |
3  | /         /// Mitigation for https://github.com/rust-lang/rust/issues/126600
4  | |         ///
5  | |         /// On `unix` (where `libc::exit` may not be thread-safe), ensure that only one Rust thread
6  | |         /// calls `libc::exit` (or returns from `main`) by calling this function before calling
...  |
18 | |         /// * If it is called again on a different thread, it will `thread::park()` in a loop
19 | |         ///   (waiting for the process to exit).
   | |________________________________________________^ help: use an automatic link instead: `<https://github.com/rust-lang/rust/issues/126600>`
   |
   = note: bare URLs are not automatically turned into clickable links
   = note: `-D rustdoc::bare-urls` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(rustdoc::bare_urls)]`
error: could not document `std`
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:00:19
  local time: Fri Jun 21 03:08:18 UTC 2024
```

</details>

zachs18 committed:
fix rustdoc URL

zachs18 commented:
Latest push moves `unique_thread_exit` to a new `std::sys::common::exit_guard` module, and just `cfg`s the definition of the function to ~~`cfg(unix)`~~ `cfg(target_os = "linux")`[^1], with an empty stub impl on other platforms[^2] (Thanks @ChrisDenton), and uses `libc::gettid`[^1] instead of `ThreadId::id()` (Thanks @the8472).

It seems like it'll pass CI (most recent failure was just doc-comment formatting issue), so I'll mark it as ready.

[^1]: `libc::gettid` appears to be linux specific, so I restricted this PR to just implementing for `target_os = "linux"`.

[^2]: other platforms can have implementations added in the `cfg_if` ladder if needed.

zachs18 committed:
Don't perform mitigation for thread-unsafe libc::exit under Miri.

1. Miri's exit is thread-safe
2. Miri doesn't (yet) support `libc::gettid`, used in the implementation of the mitigation on Linux.

zachs18 commented:
I used `core::panicking::panic_nounwind` in the "`exit` called again on the same thread" case, but considering that any `atexit` handler would necessarily be an `extern "C"` boundary (where panicking across would abort anyway IIRC), maybe this should just be a normal panic instead of an abort/non-unwinding-panic?

tbu- commented:
> Latest push moves `unique_thread_exit` to a new `std::sys::common::exit_guard` module, and just `cfg`s the definition of the function to ~`cfg(unix)`~ `cfg(target_os = "linux")`

Perhaps you could use `pthread_self` instead? It returns `pthread_t` which is an integer or pointer type on all platforms supported by `libc` and it's marked as an "async-signal-safe" function, so it'll likely work even in an exit handler.

tbu- commented:
> I used `core::panicking::panic_nounwind` in the "`exit` called again on the same thread" case, but considering that any `atexit` handler would necessarily be an `extern "C"` boundary (where panicking across would abort anyway IIRC), maybe this should just be a normal panic instead of an abort/non-unwinding-panic?

I don't think this would be a good idea, otherwise people might start trying to catch that panic (inside the `atexit` handler) for whatever reason.

RalfJung commented on library/std/src/sys/pal/common/exit_guard.rs:
```diff
@@ -0,0 +1,97 @@
+        /// This mitigation is not necessary when running under Miri, so this function does nothing
+        /// when running under Miri.
+        pub(crate) fn unique_thread_exit() {
+            // Mitigation not required on Miri, where `exit` is thread-safe.
```
```suggestion
            // Mitigation not required on Miri, where `atexit` is not supported and
            // hence `exit` is thread-safe.
```

joboet reviewed (commented):
Good thinking! I look forward to the day where this is fixed in all libcs we support...

zachs18 commented on library/std/src/sys/pal/common/exit_guard.rs:
```diff
@@ -0,0 +1,98 @@
+cfg_if::cfg_if! {
+    if #[cfg(miri)] {
+        /// Mitigation for <https://github.com/rust-lang/rust/issues/126600>
+        ///
+        /// This mitigation is not necessary when running under Miri, so this function does nothing
+        /// when running under Miri.
+        pub(crate) fn unique_thread_exit() {
+            // Mitigation not required on Miri, where `atexit` is not supported and
+            // hence `exit` is thread-safe.
+        }
```
`libc::gettid` is not usable under Miri IIUC, and [`std::thread::current().id().as_u64()` can fail](https://github.com/rust-lang/rust/pull/126606#issuecomment-2176784818) if we are currently exiting (or in  TLS destructor? (etc?)).

Is there a simple way to get a unique thread ID that is both supoorted under Miri, and works at arbitrary times? I suppose a manual atomic counter and thread_local ID that incrments the counter if not yet set would work.

 Unless it is acceptable that calling `exit` at  time when `std::thread::try_current()` fails will abort, but that would then happen even on calls to `exit` that would not otherwise abort.

RalfJung replied:
Implementing `gettid` in Miri should not be too hard. But maybe that should not block this PR.

tbu- replied:
@zachs18 Does `pthread_self` work in Miri?

RalfJung replied:
Yes, we support `pthread_self` in Miri.

zachs18 replied:
Latest push uses `pthread_self` and `Mutex<Option<pthread_t>>` instead of `libc::gettid` and `AtomicI32` for maximum portability[^1]. This also removes the code duplication since there is no "fallback" implementation. Given that this code should not be hot, I think it should be fine to not use atomics here.

One possible issue is that *formally*, comparing `pthread_t`s is supposed to use [`pthread_equal`](https://man7.org/linux/man-pages/man3/pthread_equal.3.html), not `==` comparison, but the `libc` crate currently does not expose `pthread_equal`. I don't know if any supported libcs *actually* make `pthread_equal` do anything  other than `==`, though, so it might just be fine (glibc and musl both appear to implement `pthread_equal` as `==` (glibc) ([musl](https://github.com/bminor/musl/blob/ab31e9d6a0fa7c5c408856c89df2dfb12c344039/include/pthread.h#L92))). Alternately, we could just have a one-off `extern "C" { fn pthread_equal(...) -> ...; }` above the call, at least until/unless `libc` exposes `pthread_equal`.

[^1]:  since `pthread_t` is different on different Linux platforms: `*-unknown-linux-gnu` appears to have it as `c_ulong` (which is `u64` on 64-bit and `u32` on 32-bit), `*-unknown-linux-musl` appears to have it as `*mut c_void`, and `*-linux-android` appears to have it as `c_long`

tbu- commented on library/std/src/sys/pal/common/exit_guard.rs:
```diff
@@ -0,0 +1,98 @@
+                    // Park until the process exits.
+                    drop(exiting_thread_id);
+                    loop {
+                        crate::thread::park();
```
There's also `libc::pause` which is basically what we need here. It sleeps forever.

joboet replied:
That sounds perfect, let's use that!

zachs18 replied:
Latest push uses `libc::pause()` with a `// Safety` comment stating that `libc::pause` is always safe to call, which I assume is true?

GrigorenkoPV replied:
> There's also `libc::pause` which is basically what we need here. It sleeps forever.

I am not sure it does.

From `pause(2)`[^2]:
 
> `pause()` causes  the calling process (or thread) to sleep until a signal is delivered that either terminates the process or causes the invocation of a signal-catching function.
> `pause()` returns only when a signal was caught and the signal-catching function returned.  In this case, `pause()` returns -1, `errno` is set to `EINTR`.

From `pause(3p)`[^3p]:

> The `pause()` function shall suspend the calling thread until delivery of a signal whose action is either to execute a signal-catching function or to terminate the process.
> If the action is to terminate the process, `pause()` shall not return.
> If the action is to execute a signal-catching function, `pause()` shall return after the signal-catching function returns.


[^2]: https://www.man7.org/linux/man-pages/man2/pause.2.html
[^3p]: https://www.man7.org/linux/man-pages/man3/pause.3p.html

tbu- replied:
This is not specific to the `pause` syscall, `sleep` also wakes up from signals AFAIK.

That's why it's placed in a loop.

zachs18 committed:
Use libc::pause instead of std::thread::park in wait-for-exit loop

zachs18 committed:
Use pthread_t instead of numeric thread id

zachs18 committed:
Remove Miri special-case

zachs18 committed:
Move unique_thread_exit call to lang_start_internal so it is not in a generic function, and wrap it in `catch_unwind`

zachs18 commented on library/std/src/sys/pal/common/exit_guard.rs:
```diff
@@ -0,0 +1,98 @@
+    } else if #[cfg(target_os = "linux")] {
+        /// Mitigation for <https://github.com/rust-lang/rust/issues/126600>
+        ///
+        /// On `unix` (where `libc::exit` may not be thread-safe), ensure that only one Rust thread
```
I originally had this when the PR was for `cfg(unix)`, not just `cfg(target_os = "linux")`. Latest push changes this to "UNIX-like platforms", does that sound okay?

joboet replied:
Yeah, that sounds good.

joboet reviewed (approved):
Looks fine!
r=me after moving the file and adding the comment about this being a bug in glibc.

RalfJung commented on library/std/src/sys/pal/common/exit_guard.rs:
```diff
@@ -0,0 +1,59 @@
+    if #[cfg(target_os = "linux")] {
+        /// Mitigation for <https://github.com/rust-lang/rust/issues/126600>
+        ///
+        /// On UNIX-like platforms (where `libc::exit` may not be thread-safe), ensure that only one
+        /// Rust thread calls `libc::exit` (or returns from `main`) by calling this function before
+        /// calling `libc::exit` (or returning from `main`).
+        ///
+        /// Technically not enough to ensure soundness, since other code directly calling
+        /// libc::exit will still race with this.
```
```suggestion
        /// On glibc, `libc::exit` has been observed to not always be thread-safe.
        /// It is currently unclear whether that is a glibc bug or allowed by the standard.
        /// To mitigate this problem, we ensure that only one
        /// Rust thread calls `libc::exit` (or returns from `main`) by calling this function before
        /// calling `libc::exit` (or returning from `main`).
        ///
        /// Technically, this is not enough to ensure soundness, since other code directly calling
        /// `libc::exit` will still race with this.
```

joboet replied:
I'd add the issue number #126600 as well.

zachs18 replied:
The issue is linked above this on line 3, is that good?

zachs18 committed:
Update library/std/src/sys/pal/common/exit_guard.rs

Co-authored-by: Ralf Jung <post@ralfj.de>

zachs18 committed:
Move exit guard from sys::common::exit_guard to sys::exit_guard.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9814030507/job/27101071745) [(plain)](https://github.com/rust-lang/rust/commit/a6093701437be909614bc49d63406dfaa63d3b13/checks/27101071745/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

zachs18 committed:
Attempt to fix CI

joboet commented on library/std/src/sys/mod.rs:
```diff
@@ -3,6 +3,7 @@
 /// descriptors.
 mod pal;
 
+pub(crate) mod exit_guard;
```
Please put this together with the other module declarations below. And this can be `pub`, `sys` is `pub(crate)` already.

zachs18 committed:
Move/change declaration of `mod exit_guard;`

joboet reviewed (approved):
Thank you, it's good to get this mitigated!
@bors r+

joboet commented:
@bors r+

bors commented:
:pushpin: Commit 98010765f949307bf53ebb692e45e428ba318aa1 has been approved by `joboet`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=joboet 98010765f949307bf53ebb692e45e428ba318aa1 -->
<!-- homu: {"type":"Approved","sha":"98010765f949307bf53ebb692e45e428ba318aa1","approver":"joboet","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 98010765f949307bf53ebb692e45e428ba318aa1 with merge ccb526eb82d0232a974dc66d218553b7c34a8740...
<!-- homu: {"type":"BuildStarted","head_sha":"98010765f949307bf53ebb692e45e428ba318aa1","merge_sha":"ccb526eb82d0232a974dc66d218553b7c34a8740"} -->

rust-log-analyzer commented:
The job **`dist-x86_64-musl`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/9840309613/job/27164285323) [(plain)](https://github.com/rust-lang-ci/rust/commit/ccb526eb82d0232a974dc66d218553b7c34a8740/checks/27164285323/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] object test:false 5.493
error[E0277]: `*mut libc::c_void` cannot be sent between threads safely
   --> library/std/src/sys/exit_guard.rs:26:39
    |
26  |             static EXITING_THREAD_ID: Mutex<Option<libc::pthread_t>> = Mutex::new(None);
    |                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `*mut libc::c_void` cannot be sent between threads safely
    |
    = help: within `core::option::Option<*mut libc::c_void>`, the trait `core::marker::Send` is not implemented for `*mut libc::c_void`, which is required by `sync::mutex::Mutex<core::option::Option<*mut libc::c_void>>: core::marker::Sync`
note: required because it appears within the type `core::option::Option<*mut libc::c_void>`
    |
574 | pub enum Option<T> {
    |          ^^^^^^
    |          ^^^^^^
note: required for `sync::mutex::Mutex<core::option::Option<*mut libc::c_void>>` to implement `core::marker::Sync`
    |
    |
189 | unsafe impl<T: ?Sized + Send> Sync for Mutex<T> {}
    |                         |
    |                         unsatisfied trait bound introduced here
    = note: shared static variables must have a type that implements `Sync`

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9840309613/job/27164763352)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/9840309613/job/27164763352","builder_name":"checks-actions"} -->

zachs18 committed:
Attempt to fix CI

zachs18 commented:
Latest commit should fix building on Musl (where `pthread_t` is a pointer, and thus must be wrapped to put in a `static`)

joboet commented:
That should work, let's try it.
@bors r+

bors commented:
:pushpin: Commit 8bcbab5dd1c6449c2acd3ea6a94b245c0936722d has been approved by `joboet`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=joboet 8bcbab5dd1c6449c2acd3ea6a94b245c0936722d -->
<!-- homu: {"type":"Approved","sha":"8bcbab5dd1c6449c2acd3ea6a94b245c0936722d","approver":"joboet","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

workingjubilee commented:
I have a feeling we're going to see this friend of ours again.

@bors rollup=iffy

bors commented:
:hourglass: Testing commit 8bcbab5dd1c6449c2acd3ea6a94b245c0936722d with merge 03c2100dede5d8779a032cf1fc1c28201c37048a...
<!-- homu: {"type":"BuildStarted","head_sha":"8bcbab5dd1c6449c2acd3ea6a94b245c0936722d","merge_sha":"03c2100dede5d8779a032cf1fc1c28201c37048a"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9916149579/job/27400045404)
Approved by: joboet
Pushing 03c2100dede5d8779a032cf1fc1c28201c37048a to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"joboet","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9916149579/job/27400045404"},"merge_sha":"03c2100dede5d8779a032cf1fc1c28201c37048a"} -->

rust-timer commented:
Finished benchmarking commit (03c2100dede5d8779a032cf1fc1c28201c37048a): [comparison URL](https://perf.rust-lang.org/compare.html?start=336e89bd1518f2cd81574384037b3cffe03c63a8&end=03c2100dede5d8779a032cf1fc1c28201c37048a&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=336e89bd1518f2cd81574384037b3cffe03c63a8&end=03c2100dede5d8779a032cf1fc1c28201c37048a&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.3% | [-0.3%, -0.3%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=336e89bd1518f2cd81574384037b3cffe03c63a8&end=03c2100dede5d8779a032cf1fc1c28201c37048a&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=336e89bd1518f2cd81574384037b3cffe03c63a8&end=03c2100dede5d8779a032cf1fc1c28201c37048a&stat=cycles:u)
<details>
<summary>Results (secondary -3.7%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.7% | [-4.1%, -3.2%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=336e89bd1518f2cd81574384037b3cffe03c63a8&end=03c2100dede5d8779a032cf1fc1c28201c37048a&stat=size:linked_artifact)
<details>
<summary>Results (primary 0.1%, secondary 0.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.1% | [0.0%, 0.2%] | 9     |
| Regressions ❌ <br /> (secondary)  | 0.2% | [0.0%, 0.2%] | 39    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.1% | [0.0%, 0.2%] | 9     |
</details>

**Bootstrap**: 704.601s -> 706.568s (0.28%)
**Artifact size**: 328.70 MiB -> 328.69 MiB (-0.00%)
